File: node/browserstack.js
var webdriver = require('selenium-webdriver');

exports.Node = function(framework){
  var build = webdriver.Builder.prototype.build;

  webdriver.Builder.prototype.build = function(){
    if(process.env.RUN_ON_BS){
      if(framework) this.capabilities_.set("browserstack.framework", framework);
      this.capabilities_.set("browserstack.user", process.env.BROWSERSTACK_USER);
      this.capabilities_.set("browserstack.key", process.env.BROWSERSTACK_ACCESS_KEY);
      this.url_ = "http://hub.browserstack.com:80/wd/hub";
    }
    return build.call(this);
  }
}

exports.Protractor = function(){
  var protractor = require('protractor');

  exports.Node('protractor');
}

exports.Nightwatch = function(){
  var nightwatch = require('nightwatch');

  var client = nightwatch.client;
  nightwatch.client = function(options){
    if(process.env.RUN_ON_BS){
      options["desiredCapabilities"] = {
        "browserstack.user": process.env.BROWSERSTACK_USER,
        "browserstack.key": process.env.BROWSERSTACK_ACCESS_KEY,
        "browserstack.framework": "nightwatch"
      }
      options["seleniumHost"] = "hub.browserstack.com";
      options["seleniumPort"] = 80;
    }
    return client(options);
  }

  exports.Node('nightwatch');
}

File: ruby/browserstack.rb
require 'selenium/webdriver'

module Selenium
  module WebDriver
    class << self
      alias_method :original_for, :for

      def self.for(*args)
        browser = args.shift
        opts = args.shift || {}
        opts[:url] = opts[:url] || "http://127.0.0.1:4444/wd/hub"
        opts[:desired_capabilities] = opts[:desired_capabilities] || {}
        opts[:desired_capabilities][:browserName] = browser

        if ENV["RUN_ON_BS"]
          opts[:url] = "http://#{ENV["BROWSERSTACK_USER"]}:#{ENV["BROWSERSTACK_ACCESS_KEY"]}@hub.browserstack.com/wd/hub"
          opts[:desired_capabilities]["browserstack.framework"] = BrowserStack::get_framework
        end

        original_for(:remote, opts, *args)
      end
    end
  end # WebDriver
end # Selenium

module BrowserStack
  @@framework = "ruby"

  def self.for(framework)
    @@framework = framework
  end

  def self.get_framework
    @@framework
  end
end # Browserstack
